From 35cdc81c729300613c4fe88c2622d7807c7320d3 Mon Sep 17 00:00:00 2001
From: Quentin Perret <qperret@google.com>
Date: Fri, 30 Oct 2020 13:51:18 +0000
Subject: [PATCH 798/830] ANDROID: arm64: gki_defconfig: Disable
 RANDOMIZE_MODULE_REGION_FULL
X-NVConfidentiality: public

By default, enabling RANDOMIZE_BASE on arm64 randomizes not just the
kernel base address, but also the module region independently from the
kernel, over a 4GiB range. As a consequence, modules can be out of
reach for direct branches from/to the kernel, hence creating the need to
resolve function calls via veneers in the module PLT.

While this is good for security, this also introduces a non-negligible
overhead in some benchmarks when parts of the scheduler policy is
factored out in modules, as is the cases with GKI 2.0 on some devices.

To get the performance back, disable RANDOMIZE_MODULE_REGION_FULL, which
limits the range of randomization for the module area to remain within
reach of direct branches. This is clearly a security trade-off as the
module and kernel addresses have a stronger correlation, but this still
remains preferable over the alternative which is no-GKI and having all
drivers compiled into the kernel itself, as Android used to be in
previous generations.

Bug: 168521642
Signed-off-by: Quentin Perret <qperret@google.com>
Change-Id: I8f4e05fbd308f99f38a4a2b33802ef1e195666e3
---
 arch/arm64/configs/gki_defconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/arm64/configs/gki_defconfig b/arch/arm64/configs/gki_defconfig
index c83ace1982ac..ddbaee9aed53 100644
--- a/arch/arm64/configs/gki_defconfig
+++ b/arch/arm64/configs/gki_defconfig
@@ -58,6 +58,7 @@ CONFIG_SWP_EMULATION=y
 CONFIG_CP15_BARRIER_EMULATION=y
 CONFIG_SETEND_EMULATION=y
 CONFIG_RANDOMIZE_BASE=y
+# CONFIG_RANDOMIZE_MODULE_REGION_FULL is not set
 # CONFIG_DMI is not set
 CONFIG_PM_WAKELOCKS=y
 CONFIG_PM_WAKELOCKS_LIMIT=0
-- 
2.17.1

